storage/replica.go

```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
+		// also need the Batch.Defer mechanism.
 		var err error
 		if lastIndex, err = r.applySnapshot(batch, rd.Snapshot); err != nil {
 			return err
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
+	// We know that all of the writes from here forward will be to distinct keys.
```

Comment:
My recollection is that `distinctBatch` has its current semantics so that you didn't have to recall when you created it to know what writes it would see. That is simple at one level, but as you've noted it adds cognitive complexity at another. Let me take a look at making the change you suggest. I don't think it will be difficult.